docs/design/sql/16-nested-fields-access.md

```diff
@@ -0,0 +1,359 @@
+
+The Predicate API, which SQL is aiming to replace, allowed access to nested
+fields of the keys or values. SQL so far lacks this feature.
+
+Hazelcast is most commonly used by Java shops, which tend to store complex
+objects. This is contrary to traditional RDBMS, where denormalized structures
+are typically used. This made the feature very useful for our user base.
+
+User Defined Types are meant to provide users with ability to query nested
+fields inside their OBJECT mappings, by providing both SQL Backend and
+Frontend with necessary type information required for validation and execution
+of queries.
+
+The original motivation was to support queries like `SELECT
+(organization).office.name FROM users` where `organization` would be an OBJECT
+field itself (e.g. a Java class or a nested GenericRecord) with its own
+columns/fields (here and thereafter Column and Field are used interchangeably).
+This was also expanded to include ability to perform UPDATE/INSERT queries as
+well as support for Portable and Compact MAPPINGs in addition to Java ones.
+
+## Terminology
+
+|Term|Definition|
+|---|---|
+|||
+|UDT|**User Defined Types**, also known as Custom Types, also previously known as Nested Type and Nested Fields Types|
+|CRec Type| Cyclically Recurrent UDTs, UDT type hierarchies with type or instance-level cycles|
+|Nested Field|A column (field) inside of mapped OBJECT column (field)|
+|Nested Type| In this document, synonymous to UDT|
+
+## Goals
+
+The feature should provide:
+
+- using nested fields to filter rows (WHERE clause and JOIN conditions)
+- using nested fields in the results and for aggregation
+- updating fields of complex objects in the UPDATE statement
+- inserting complex objects by providing values to all fields
+
+## Non-goals
+
+- Provide access to polymorphic data (i.e. a value has multiple subclasses, the
+  user will be able to use the various fields of different subclasses)
+
+## Structured vs. non-structured data
+
+We already support non-structured JSON data using the JSON family of functions.
+In this effort we're aiming at structured data, that is those, whose schema is
+known in advance. That's why we'll need the custom types to be created before
+they can be used.
+
+# Functional Design
+
+## Current status
+
+SQL currently supports the `OBJECT` type. This type is used for all fields that
+don't have one of the more specific type (we call them SQL primitive types). 
+
+Behavior of `OBJECT`:
+
+- supported operators: comparison operators (`=`, `<`, `>`, ...) (all use
+  `compareTo`, supported only for instances the cluster can deserialize to java
+  instances, doesn't work for `GenericRecord`), `IS NULL` is also supported
+
+- can be cast to `VARCHAR` (uses `toString()` on the java instance or on
+  `GenericRecord`)
+
+- when returned to client, either Java instance or, if not available, a
+  `GenericRecord` is returned. They are only supported in java client, other
+  clients will fail (TODO confirm)
```

Comment:
This would not be possible right now most likely. There's no built in mechanism for this (tested on a client-server bench with different classpaths) and its most likely going to be challenging to implement this since this would require server to check with client whether it has given class in its classpath and then send it. Additionally sending objects with cycles would not be possible and we'd have to break the cycles to convert the data to GenericRecords.

Overall I'd say this isn't feasible to implement in 5.2.